### ISSUE TITLE
Letting 0 be a valid value

### ISSUE THREAD

**Issue by garengllc** [deepsig/libsigmf#3]
I probably missed this somewhere, but I have a custom namespace with a field test in the global object.  The value of 0 (a int) is valid, but it doesn't show up when I print the JSON value.  I think that it is being ignored as a blank field. Is there a way to force this to be printed?

<<EDIT>> I believe this to be a JSON issue, so I am closing it.

**Comment by garengllc** [comment]
OK, I think there might be more to this issue than I thought.  I believe there isn't a problem with the JSON, so that leads me back to this library.
If I change the example `example_sigmf_json_roundtrip.cpp` to have the "one fish" sample_cout to be 0 instead of 1, it doesn't print it (as to be expected based on my issue I am seeing), but it also segfaults:
`example_sigmf_json_roundtrip: /home/jason//SigMF/src/libsigmf/examples/example_sigmf_json_roundtrip.cpp:61: int main(): Assertion `back_to_json.dump(4) == starting_str' failed.`, which is surprising to me as I would have thought that it would just throw an error on the `assert(back_to_json.dump(4) == starting_str);`, not segfault.

**Comment by garengllc** [comment]
OK, I found the area causing me issues.  In `flatbuffers_json_visitor.h`, the code block:
```
        if (val) {
            if (ref_idx >= 0) {
                ttptr = typetable->type_refs[ref_idx]();
            }
            if (is_vector) {
                val += flatbuffers::ReadScalar<flatbuffers::uoffset_t>(val);
                auto vec = reinterpret_cast<const flatbuffers::Vector<uint8_t> *>(val);
                auto elem_ptr = vec->Data();
                for (size_t j = 0; j < vec->size(); j++) {
                    json_object[ns_prefix + name].push_back(
                            flatbuffer_field_to_json(elem_ptr, type, ttptr, ns_prefix, include_defaults));
                    elem_ptr += InlineSize(type, ref);
                }
            } else {
                json lval = flatbuffer_field_to_json(val, type, ttptr, ns_prefix, include_defaults);
                json_object[ns_prefix + name] = lval;
            }
        } else {
            // There is no value in this field. This is OK except maybe we want to figure out how to enforce
            // required fields. Although maybe flatbuffers schema language and the flatbuffersbuilders have
            // already enforced such things.
        }
```
check if there is  value.  It isn't seeing a value for the items that I set to 0, it is falling into the else statement at the bottom, where you commented on what is causing me issues.  I tied it to if(true) at the top and it not prints fine.  This obviously is not desired, but it is a cheap workaround with minimal code changes.

**Comment by n-west** [comment]
Hi, yea... this is an annoying issue. The root of the problem is basically using flatbuffers as a code generator & tree parser. If a value has the default value then it doesn't get parsed well by this tree and there's not a good way to say "this field specifically is required". What we have done so far is used the flag for "print all" to include all default values

**Comment by garengllc** [comment]
It makes sense, and I don't know what a good answer would be.

I know I am being dumb, but I couldn't find the flag you were referring to.

**Comment by n-west** [comment]
There's the option to the to_json member
```
        json to_json(bool include_defaults=false) const {
            constexpr auto tuple_size = std::tuple_size<sigmftypes>::value;
            constexpr auto at_the_end = ReachedEnd<0, tuple_size>::value;
            json r = TupleIterator<0, sigmftypes, at_the_end>::to_json(sigmf_namespaces, include_defaults);
            return r;
        }
```

**Comment by jacobagilbert** [comment]
@garengllc not sure if this is still relevant for you, but libsigmf 1.0 will switch to making scalars truly optional. See:

https://github.com/google/flatbuffers/pull/6270
https://github.com/google/flatbuffers/issues/6014
https://github.com/google/flatbuffers/pull/6092
https://github.com/google/flatbuffers/pull/6155

There is a v1.0 branch on this repository now, though it will almost certainly change before release.

**Comment by jacobagilbert** [comment]
This has been fully implemented with the `v1.0.1` release.
